### PR TITLE
05875 query workflow improvements part3

### PR DESCRIPTION
**Pull request by netopyr**
Next and hopefully final part of changes that improve query and ingest workflow

- Added facility `QueryFeeCheck`
- Added check if payer can pay for fees
- Removed `Counter` from `IngestWorkflowImpl`

### CODE REVIEW

**Review by rbair23 — APPROVED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java [hunk 1]
```diff
@@ -162,21 +161,6 @@ public ResponseCodeEnum validateQueryPaymentTransfers(final TransactionBody txn)
         return OK;
     }
 
-    public com.hedera.hapi.node.base.ResponseCodeEnum validateQueryPaymentTransfers2(
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java:165]: Does it makes sense to deprecate this class?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Implementation of {@link QueryFeeCheck} that is based on the mono-service implementation.
+ */
+public class MonoQueryFeeCheck implements QueryFeeCheck {
+
+    private final com.hedera.node.app.service.mono.queries.validation.QueryFeeCheck delegate;
+
+    @Inject
+    public MonoQueryFeeCheck(com.hedera.node.app.service.mono.queries.validation.QueryFeeCheck delegate) {
+        this.delegate = delegate;
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java:38]: I would do null checks here, too.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        final var monoTxBody = PbjConverter.fromPbj(txBody);
+        final var monoResult = delegate.validateQueryPaymentTransfers(monoTxBody);
+        final var result = PbjConverter.toPbj(monoResult);
+        if (result != OK) {
+            throw new InsufficientBalanceException(result, queryFee);
+        }
+    }
+
+    @Override
+    public void nodePaymentValidity(List<AccountAmount> transfers, long queryFee, AccountID node)
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java:53]: null check for transfers

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    private final com.hedera.node.app.service.mono.queries.validation.QueryFeeCheck delegate;
+
+    @Inject
+    public MonoQueryFeeCheck(com.hedera.node.app.service.mono.queries.validation.QueryFeeCheck delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void validateQueryPaymentTransfers(TransactionBody txBody, long queryFee)
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java:42]: null annotations

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/QueryFeeCheck.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Validates query payment transfer transaction before reaching consensus. Validate each payer has enough balance
+     * that is needed for transfer. If one of the payer for query is also paying transactionFee validate the payer has
+     * balance to pay both
+     *
+     * @param txBody the transaction body to validate
+     * @param queryFee transaction fee
+     * @throws InsufficientBalanceException if the transaction is invalid
+     */
+    void validateQueryPaymentTransfers(TransactionBody txBody, long queryFee) throws InsufficientBalanceException;
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/QueryFeeCheck.java:39]: null annotations

**Review by hendrikebbers — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java (hunk 2 above)
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java]: done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java (hunk 4 above)
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java]: done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java (hunk 3 above)
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/MonoQueryFeeCheck.java]: done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/QueryFeeCheck.java (hunk 5 above)
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/QueryFeeCheck.java]: done

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java (hunk 1 above)
Comment (netopyr) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java:165]: No, the class is still being used. The mono-service uses it and hedera-app delegates its call. All I did was to extract these methods into an adapter.

Comment (hendrikebbers) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java:165]: the class is still being used <- I do not ask do delete it ;) But by adding the @Deprecated annotation and a short message other developers will know about the new facility and do not use the old classes in new / refactored code

**Review by hendrikebbers — APPROVED** [review]

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java (hunk 1 above)
Comment (netopyr) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/validation/QueryFeeCheck.java:165]: Ah, right. Sorry, misunderstood you.

It is not so much about old code vs. new code. Code in `mono-service` should continue to use this class, code in `hedera-app` should use the new class. Adding the `@Deprecated` annotation would be confusing, if you are working within mono-service.

**Review by hendrikebbers — DISMISSED** [review]

**Review by mhess-swl — APPROVED** [review]

**Review by rbair23 — APPROVED** [review]

**Pull request merged**